### ISSUE TITLE
RFC: Release process, cycles, scheduling and versioning

### ISSUE THREAD

**Issue by hifi** [keepassxreboot/keepassxc#7237]
Our current release process is not well defined and we don't have a release schedule. This has some downsides and I'll try to document a model that could work for us.

To start off I've noticed at least the following issues:
 - releases have no schedule
 - our versioning is stupid
 - spacing between releases have become very long
 - hard to decide the point when to release

In this document/issue I describe a process that in my opinion would fit our model but with minimal changes to what we currently do.

Scheduling
----------
To start off I propose we have a defined timed release cycle that ensures we have a steady output that is easy for contributors and packagers to predict. This makes it easy for everyone to know when a PR is merged when a release with the merged code would be expected. Packagers will know when to expect a release so they may be more alert to swiftly do their part.

Our schedule should define the exact end-of-month when we are going to start our release process and we should not deviate from that or stall it because there's a nice PR waiting. Because we have a defined schedule it doesn't take years for the release with that nice PR to land.

We don't start the release process if we have nothing to release. The cycle continues up to next release point in this case. Bug fixes are released between cycles so this would only affect when we have no features merged in.

Versioning
----------
Currently our major version is frozen and we only do minor and micro versioning and it makes it extremely hard for packagers or users to understand the importance of our releases right now.

We should definitely do semantic versioning. Major for big features and breaking changes, minor for smaller backwards compatible improvements and micro for bug fixes only. People understand semantic versioning so it doesn't need too much explaining either. We shouldn't be afraid to bump our major version.

The next feature version (major or minor) is determined during a release cycle depending what has been merged in. We don't do add any features into micro versions, only fix bugs.

Release process
---------------
We continue to merge PRs into develop like we currently do but when we are approaching the _end_ of a release cycle we branch off a versioned release branch. This will be our next feature release.

Immediately after branching we release RC0 from the branch. This is the point when no more features will be merged into this release and only bug fixes will land in the release branch and more RCs released. Large scale testing shall commence.

Our RC process must have a clear timed schedule as well. Each RC would be in a testing phase for as long as we have defined and if no release blocking bugs are found we release that as-is without second guessing.

After a feature release is made there are usually bugs we or anyone testing an RC (or recent snapshot) have not noticed. We branch a bug fix release from the feature release branch and start merging fixes into that. As it depends on the severity of the bugs we can keep our current style of releasing early without a defined process for that.

Between all this and the original branching off from develop we can continue to merge things into develop like we normally do. We only merge the fixes from the release branches back to develop. The master branch can point to our latest release like it has.

Comments
--------
All this doesn't really add much more work for us but what it does is that we need to be able to hold off merging things that are not ready and try to keep our review process quality high so the next RC will not have too many broken things. We stabilize things within the RC process. This is pretty much how the Linux kernel releases work.

We only work towards a single next release and we only fix bugs for our latest feature release so we don't need to think about more than what we do right now - our current release and develop.

The biggest change is we _just do it_ when it's time and not wait for some magical moment when there's enough features baked in as it's hard to determine and we always want _that one thing_ more. We could have released at least two feature releases before 2.7 with a process like this and releases would be much smaller and get more testing (in production) earlier so we don't end up with too many unique bugs per feature release.

Questions
---------
0. We cool with semantic versioning?
1. How long our feature release cycles should be: 6 months, 12 months, something else?
2. How long does an RC need to be out without release blocking bugs? 2 weeks, 3 weeks, something else?
3. Would our release quality be affected by this process by making them less stable or secure?
4. Can people who work as "release managers" be bothered to take the time to do it when expected?

**Comment by droidmonkey** [comment]
This needs to happen. Release early, release often. Here are my timing suggestion:

* First digit - once per year, release month planned in advance
* Second digit - twice per (April and August)
* Third digit - as needed to maintain stability or fix minor things in current baseline. 

2.7.0 will be our second digit January release. 

**What goes into a release?**

First digit changes should start with long term planning to pick 3-4 major changes (mainly UX/UI or format based) to be the focus of the release. Beta releases should be out for a month prior to final.

Second digit releases will introduce new features, but avoid major code refactors and dependency changes. Major UX/UI changes should also be avoided. Release candidates should be out two weeks prior to final.

Third digit changes will be reserved for minor bug fixes, security fixes, and minor UI tweaks that fix annoyances or bugs. These are released directly to final. 

**Using Git**

First digit release work should be placed in a separate release branch off of develop to avoid major refactors from inhibiting second digit progress. The first digit branch can be rebased onto develop as needed. 

Second digit work should be merged directly to develop. Release branch made for the release candidate. 

Third digit work should be merged to develop and cherry picked onto a release branch off master. 

**Using GitHub**

We should use GitHub projects to track second and first digit work and release burn down.

Use labels to mark issues for their complexity to help include them into the proper release planning/cycle. We need to do better at marking issues for first time / drive by contributors.

Our documentation needs an overhaul to help people get into the project and contributing as best they can.

**Comment by hifi** [comment]
I'm afraid that's too many scheduled releases to handle in a year as that would require almost a release per month. I have no problem with the other proposed ways of the process including not using semantic versioning.

I'd just release "what we have" and decide the version bump based on that and have more of those scheduled like three or four per year excluding third digit which is on demand. Keeping the big refactors in separate branches can still work and just merged when they are ready so they will be picked into the next release.

**Comment by varjolintu** [comment]
Concering droidmonkey's message, I'd prefer that if there's a one major release per year, the second digit minor releases should be twice in a year. If a major release would be in (for example) December, the minor ones could be in April and August. Otherwise there's three minor releases in a year plus some major release in somewhere between?

Third digit releases sound fine. Bug releases etc. should be released more often, like every other month as proposed.

**Comment by droidmonkey** [comment]
I agree with the above and modified my original response.  I do think we need to curate the big refactors and coordinate their release. I'm not a fan of releasing those "whenever they are ready". As we've seen with 2.7.0 a lot of changes relied on the previous ones (Botan, then MSVC, then remove quazip dependency, then Windows Hello, etc).
